File: aerosync-backend/src/main/java/com/aerosync/service/WeatherService.java
package com.aerosync.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.time.LocalDateTime;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Slf4j
@Service
public class WeatherService {

    @Value("${weather.api.key:}")
    private String weatherApiKey;

    @Value("${weather.api.url:https://api.openweathermap.org/data/2.5}")
    private String weatherApiUrl;

    private final ObjectMapper objectMapper = new ObjectMapper();

    // Cache: airport code → {score, timestamp}
    private final Map<String, CachedWeather> cache =
            new ConcurrentHashMap<>();

    private static final long CACHE_TTL_MS = 30 * 60 * 1000; // 30 mins

    // Indian airport IATA → city for OpenWeatherMap
    private static final Map<String, String> AIRPORT_CITIES = Map.of(
        "BOM", "Mumbai,IN",
        "DEL", "Delhi,IN",
        "BLR", "Bangalore,IN",
        "HYD", "Hyderabad,IN",
        "CCU", "Kolkata,IN",
        "MAA", "Chennai,IN",
        "AMD", "Ahmedabad,IN",
        "GOI", "Goa,IN"
    );

    // ── Get weather score for airport ─────────────────────────────────────────
    // Returns 0-3 score. Cached for 30 minutes.
    // Falls back to 1 if API unavailable.
    public int getWeatherScore(String airportCode) {
        if (weatherApiKey == null || weatherApiKey.isBlank()) {
            log.debug("Weather API key not set — using default score");
            return 1;
        }

        // Check cache
        CachedWeather cached = cache.get(airportCode);
        if (cached != null
                && (System.currentTimeMillis() - cached.timestamp)
                   < CACHE_TTL_MS) {
            log.debug("Weather cache hit for {}: score={}",
                    airportCode, cached.score);
            return cached.score;
        }

        // Fetch from API
        try {
            String city = AIRPORT_CITIES.get(airportCode.toUpperCase());
            if (city == null) return 1;

            String url = weatherApiUrl + "/weather"
                    + "?q=" + city
                    + "&appid=" + weatherApiKey
                    + "&units=metric";

            String responseStr = WebClient.builder()
                    .build()
                    .get()
                    .uri(url)
                    .retrieve()
                    .bodyToMono(String.class)
                    .timeout(java.time.Duration.ofSeconds(5))
                    .block();

            if (responseStr == null) return 1;

            JsonNode data  = objectMapper.readTree(responseStr);
            int score      = parseWeatherScore(data);

            // Cache result
            cache.put(airportCode, new CachedWeather(
                score, System.currentTimeMillis()));

            log.info("🌤 Live weather for {}: score={} ({})",
                    airportCode, score,
                    data.path("weather").get(0)
                        .path("description").asText("unknown"));

            return score;

        } catch (Exception e) {
            log.warn("Weather fetch failed for {}: {}",
                    airportCode, e.getMessage());
            return 1; // safe default
        }
    }

    // ── Parse OpenWeatherMap response to 0-3 score ───────────────────────────
    private int parseWeatherScore(JsonNode data) {
        int    weatherId  = data.path("weather").get(0)
                               .path("id").asInt(800);
        double windSpeed  = data.path("wind")
                               .path("speed").asDouble(0);
        int    visibility = data.path("visibility").asInt(10000);
        double rain1h     = data.path("rain")
                               .path("1h").asDouble(0);

        int score = 0;

        // Thunderstorm — always severe
        if (weatherId >= 200 && weatherId <= 232) score = 3;

        // Heavy rain
        else if (weatherId == 502 || weatherId == 503
                || weatherId == 504) score = 3;

        // Moderate rain
        else if (weatherId == 501
                || (weatherId >= 311 && weatherId <= 314)) score = 2;

        // Light rain or drizzle
        else if ((weatherId >= 300 && weatherId <= 321)
                || weatherId == 500) score = 1;

        // Snow
        else if (weatherId >= 600 && weatherId <= 622)
            score = rain1h > 1 ? 2 : 1;

        // Fog — critical for airports
        else if (weatherId == 741 || weatherId == 701) score = 2;

        // High wind
        if (windSpeed > 15) score = Math.max(score, 2);
        else if (windSpeed > 10) score = Math.max(score, 1);

        // Low visibility
        if (visibility < 1000) score = Math.max(score, 3);
        else if (visibility < 3000) score = Math.max(score, 2);
        else if (visibility < 5000) score = Math.max(score, 1);

        return Math.min(score, 3);
    }

    // ── Update all flights with live weather ──────────────────────────────────
    // Called by FlightService.predictAllFlights()
    // and by scheduled job every 30 minutes
    public Map<String, Integer> getWeatherForAllAirports() {
        Map<String, Integer> scores = new ConcurrentHashMap<>();
        for (String airport : AIRPORT_CITIES.keySet()) {
            scores.put(airport, getWeatherScore(airport));
        }
        return scores;
    }

    // Simple cache record
    private record CachedWeather(int score, long timestamp) {}
}

File: aerosync-ml/services/weather_service.py
import requests
import os
from datetime import datetime

# Indian airport ICAO → city name for OpenWeatherMap
AIRPORT_CITIES = {
    "BOM": "Mumbai,IN",
    "DEL": "Delhi,IN",
    "BLR": "Bangalore,IN",
    "HYD": "Hyderabad,IN",
    "CCU": "Kolkata,IN",
    "MAA": "Chennai,IN",
    "AMD": "Ahmedabad,IN",
    "GOI": "Goa,IN",
    "PNQ": "Pune,IN",
}

# Cache weather to avoid hitting API on every prediction
# Refresh every 30 minutes
_weather_cache = {}
_cache_timestamp = {}
CACHE_TTL_SECONDS = 1800  # 30 minutes


def get_weather_score(airport_code: str,
                      api_key: str) -> dict:
    """
    Fetch real weather for airport and convert to 0-3 score.
    Uses OpenWeatherMap free tier.
    Returns cached result if fresh enough.
    """
    now = datetime.now().timestamp()
    cached_time = _cache_timestamp.get(airport_code, 0)

    # Return cached if still fresh
    if (now - cached_time) < CACHE_TTL_SECONDS \
            and airport_code in _weather_cache:
        return _weather_cache[airport_code]

    city = AIRPORT_CITIES.get(airport_code.upper())
    if not city:
        return _fallback_weather(airport_code)

    try:
        url = "https://api.openweathermap.org/data/2.5/weather"
        params = {
            "q":     city,
            "appid": api_key,
            "units": "metric"
        }
        response = requests.get(url, params=params, timeout=10)

        if response.status_code != 200:
            print(f"Weather API error for {airport_code}: "
                  f"{response.status_code}")
            return _fallback_weather(airport_code)

        data   = response.json()
        result = _parse_weather(data, airport_code)

        # Cache the result
        _weather_cache[airport_code]    = result
        _cache_timestamp[airport_code]  = now

        return result

    except Exception as e:
        print(f"Weather fetch failed for {airport_code}: {e}")
        return _fallback_weather(airport_code)


def _parse_weather(data: dict, airport_code: str) -> dict:
    """
    Convert OpenWeatherMap response to AeroSync weather score.
    Score 0 = clear, 1 = light, 2 = moderate, 3 = severe
    """
    weather_id   = data.get("weather", [{}])[0].get("id", 800)
    wind_speed   = data.get("wind", {}).get("speed", 0)     # m/s
    visibility   = data.get("visibility", 10000)              # metres
    rain_1h      = data.get("rain",  {}).get("1h", 0)        # mm
    snow_1h      = data.get("snow",  {}).get("1h", 0)        # mm
    humidity     = data.get("main",  {}).get("humidity", 50)
    description  = data.get("weather", [{}])[0].get(
        "description", "clear")

    score = 0

    # Thunderstorm (200-232) — always severe
    if 200 <= weather_id <= 232:
        score = 3

    # Heavy rain (502-504, 522) — severe
    elif weather_id in [502, 503, 504, 522]:
        score = 3

    # Moderate rain (501) or heavy drizzle — moderate
    elif weather_id in [501, 311, 312, 313, 314]:
        score = 2

    # Light rain (500) or drizzle (300-321) — light
    elif 300 <= weather_id <= 321 or weather_id == 500:
        score = 1

    # Snow (600-622)
    elif 600 <= weather_id <= 622:
        score = 2 if snow_1h > 1 else 1

    # Fog/mist (741, 701) — critical for airports
    elif weather_id in [741, 701, 711, 721]:
        score = 2

    # High wind regardless of precipitation
    if wind_speed > 15:   # > 54 km/h
        score = max(score, 2)
    elif wind_speed > 10:  # > 36 km/h
        score = max(score, 1)

    # Low visibility
    if visibility < 1000:
        score = max(score, 3)
    elif visibility < 3000:
        score = max(score, 2)
    elif visibility < 5000:
        score = max(score, 1)

    return {
        "airport_code":   airport_code,
        "weather_score":  min(score, 3),
        "description":    description,
        "wind_speed_ms":  round(wind_speed, 1),
        "visibility_m":   visibility,
        "rain_1h_mm":     rain_1h,
        "humidity_pct":   humidity,
        "raw_weather_id": weather_id,
        "is_live":        True
    }


def _fallback_weather(airport_code: str) -> dict:
    """Return safe default when API unavailable."""
    return {
        "airport_code":  airport_code,
        "weather_score": 1,
        "description":   "data unavailable",
        "is_live":       False
    }


def get_all_airports_weather(api_key: str) -> dict:
    """Fetch weather for all 4 main airports at once."""
    results = {}
    for airport in ["BOM", "DEL", "BLR", "HYD"]:
        results[airport] = get_weather_score(airport, api_key)
    return results


if __name__ == "__main__":
    # Test — replace with your key
    key = os.environ.get("WEATHER_API_KEY", "your_key_here")
    weather = get_all_airports_weather(key)
    for airport, data in weather.items():
        print(f"{airport}: score={data['weather_score']} "
              f"({data['description']})")